jobserver/project.py

```diff
@@ -77,25 +80,67 @@ def load_yaml(content):
     return yaml.safe_load(content)
 
 
-def render_definition(content, link_func):
+def map_run_scripts_to_links(content, link_func):
     """
-    Build a HTML version of the given project.yaml content
+    Build a mapping of script-looking substrings to HTML link
 
-    In the future it might be eaiser to consume a parsed version of a
-    project.yaml but since we're only convert the script-looking substrings
-    currently this seems like the quicker path.
+    We want to link scripts in project.yamls to their files on GitHub, however
+    we can't replace the script names in YAML with <a>s as pygments will wrap
+    it in its output.  So instead this function builds up a mapping with the
+    original target for a run: as the key and the same line with scripts
+    converted to links as the value.  That lets us turn a line which looks so:
+
+        run: stata-mp:latest analysis/flow_chart_af_population.do af_population_flowchart
+
+    into a mapping like this:
+
+        {"stata-mp:latest analysis/flow_chart_af_population.do af_population_flowchart": "<a href="…">stata-mp:latest analysis/flow_chart_af_population.do af_population_flowchart</a>"}
+
+    This lets us then mutate the HTML output from pygments.
     """
-    lines = content.strip().split("\n")
+    # break up the original content into lines
+    lines = content.split("\n")
 
-    for i, line in enumerate(lines):
-        if "run:" not in line:
-            continue
+    # gather only lines with a run: stanza
+    run_lines = [line.strip() for line in lines if "run:" in line]
+
+    # strip the run prefixes so the resulting line can be used as a key later
+    # the HTML generated by pygments puts the `run:` portion in a different
+    # <span> so we don't need to bother keeping it when replacing the target
+    # portion.
+    run_lines = [line.removeprefix("run:").strip() for line in run_lines]
 
-        # recreate the spaces removed inside parse_run_line()
-        lines[i] = " ".join(link_run_scripts(line, link_func))
+    # build up an iterable of tuples with (original, replacement)
+    links_map = (
+        (line, " ".join(link_run_scripts(line, link_func))) for line in run_lines
+    )
 
-    # replace newlines with <br /> elements so the normal newlines aren't
-    # collapsed when the browser renders them.
-    definition = "\n".join(lines)
+    # compact the list to remove Nones, from lines with no appropriate script
+    # as the run target (eg a generate_cohort action)
+    links_map = (m for m in links_map if m is not None)
```

Comment:
I think this line is redundant because `m` will always be a tuple now, won't it? And if a line doesn't have any scripts in it it will just get mapped to itself and so doesn't need special handling.

Reply:
Oh yes, good catch!